src/library/scala/collection/Set.scala

```diff
@@ -176,6 +176,10 @@ trait SetOps[A, +CC[_], +C <: SetOps[A, CC, C]]
   @deprecated("Use &- with an explicit collection argument instead of - with varargs", "2.13.0")
   def - (elem1: A, elem2: A, elems: A*): C = diff(elems.toSet + elem1 + elem2)
 
+  // The implicit dummy parameter is necessary to avoid erased signature clashes
+  // between this `concat` and the polymorphic one defined in `IterableOps`.
+  // Note that these clashes only happen in Dotty because it adds mixin
+  // forwarders before erasure unlike Scala 2.
```

Comment:
Would this implicit param be removed at some point in the future, i.e. in 3.X, or is it permanent? If permanent, then maybe it's better to just name this method something else?

Reply:
> Would this implicit param be removed at some point in the future, i.e. in 3.X, or is it permanent? 

Permanent.

> If permanent, then maybe it's better to just name this method something else?

There are alternative names for `concat`/`++` in Set already: `union`/`|`. But then that means we need to tell everyone to not use `concat`/`++` if they don't want to get worse performance.

> seems pretty gross to have these dummy implicits turn up in Scaladoc and so forth...

Agreed. Note that this isn't without precedent: `AnyRefMap` does the same thing for `map`, `flatMap` and `collect`. One cosmetic improvement would be to hide implicit parameter lists consisting only of DummyImplicit in scaladoc and the IDE hover-on-type.

I also considered various alternatives to this and discussed them with Julien and others but I wasn't able to come up with anything better: I can't think of a way to twist the various definitions of `concat` to not cause these signature clashes without either breaking soundness or worsening performance. And I don't want to change Dotty to add mixin forwarders after erasure like Scala 2 does, because this creates other problems (see e.g. https://github.com/scala/scala-dev/issues/178).

Reply:
To add some context: the root cause of the issue here is that we both have a `def concat[B >: A](that: IterableOnce[B]): CC[B]` in `Iterable` and a `def concat(that: IterableOnce[A]): C` in `Set`. The latter can't be in `IterableOnce` because of variance, because it knows that it takes an `IterableOnce[A]` it can be implemented more efficiently, it also returns a more precise result type (which only matters when `CC[A] != C`, which is only the case for `BitSet` and `ValueSet` I think).

Reply:
(Accidentally clicked "Resolve")

> >If permanent, then maybe it's better to just name this method something else?
> 
> There are alternative names for concat/++ in Set already: union/|. But then that means we need to tell everyone to not use concat/++ if they don't want to get worse performance.


What do we think about widening the argument of `union` to `collection.IterableOnce[A]` then, and pattern matching for case of `Set[A]`?

Reply:
You mean, on its own ? Or combined with removing the monomorphic concat ? The former is fine, the latter doesn't solve the problem of making performance worse for everyone using concat/++.

Reply:
Well I did mean the latter, but yeah I see what you mean 😞

Reply:
We could also imagine a language feature to "unclash" your method, e.g. writing `@unclash1 def foo` would make the compiler generate code equivalent to us manually adding one DummyImplicit. Library authors might appreciate something like that.

Reply:
If we're throwing out  ideas, maybe something like implicit argument of type `B <:< A` but which always is filled in by compiler and at runtime can be inspected for whether or not it is true. Like

```scala
class Foo[A](a: A) {
  def bar[B](b: B)(implicit ev: B <:<? A): Unit = {
    if (ev.isTrue) println(ev(b)) else println(a)
  }
}
```